pkg/placement/placement-decision.go

```diff
@@ -0,0 +1,144 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/kubestellar/kubestellar/api/edge/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// syncPlacementDecision syncs a placement-decision object with what is resolved by the placement-resolver.
+func (c *Controller) syncPlacementDecision(key util.Key) error {
+	obj, err := c.getObjectFromKey(key)
+	if err != nil {
+		// The resource doesn't exist, needs to be created
+		obj = util.EmptyUnstructuredObjectFromKey(key)
+	}
+
+	unstructuredObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("failed to convert runtime.Object to unstructured.Unstructured")
+	}
+
+	placementDecision, err := unstructuredObjectToPlacementDecision(unstructuredObj)
+	if err != nil {
+		return fmt.Errorf("failed to convert from Unstructured to PlacementDecision: %w", err)
+	}
+
+	// placement decision name matches that of the placement 1:1, therefore its NamespacedName is the same.
+	placementIdentifier := placementDecision.GetName()
+
+	// generate placement decision spec from resolver
+	generatedPlacementDecisionSpec, err := c.placementResolver.GeneratePlacementDecision(placementIdentifier)
+	if err != nil {
+		return fmt.Errorf("failed to get placement decision spec: %w", err)
+	}
+
+	// calculate if the resolved decision is different from the current one
+	if !c.placementResolver.ComparePlacementDecision(placementIdentifier, &placementDecision.Spec) {
+		// update the placement decision object in the cluster by updating spec
+		if err = c.updateOrCreatePlacementDecision(placementDecision, generatedPlacementDecisionSpec); err != nil {
+			return fmt.Errorf("failed to update or create placement decision: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// updateOrCreatePlacementDecision updates or creates a placement-decision object in the cluster.
+// If the object already exists, it is updated. Otherwise, it is created.
+func (c *Controller) updateOrCreatePlacementDecision(pd *v1alpha1.PlacementDecision,
+	placementDecisionSpec *v1alpha1.PlacementDecisionSpec) error {
+	// use the passed placement decision and set its spec
+	pd.Spec = *placementDecisionSpec
+
+	// set owner reference
+	//	get owner placement object
+	placement, err := c.getPlacementByName(pd.GetName())
+	if err != nil {
+		return fmt.Errorf("failed to get placement %s: %w", pd.GetName(), err)
+	}
+
+	setOwnerReference(placement, pd)
+
+	// update or create placement decision
+	unstructuredPlacementDecision, err := placementDecisionToUnstructuredObject(pd)
+	if err != nil {
+		return fmt.Errorf("failed to update or create placement decision: %w", err)
```

Comment:
This error message should be more specific about what went wrong, particularly since the caller is going to wrap with another copy of the same words ("failed to update or create placement decision").